cunumeric/array.py

```diff
@@ -3805,3 +3835,69 @@ def _perform_where(cls, mask, one, two) -> ndarray:
         out = ndarray(shape=out_shape, dtype=common_type, inputs=args)
         out._thunk.where(mask, one, two)
         return out
+
+    @classmethod
+    def _perform_scan(
+        cls, op, src, axis=None, dtype=None, out=None, nan0=False
+    ) -> ndarray:
+        if dtype is None:
+            if out is None:
+                if np.issubdtype(src.dtype, np.integer):
+                    # Set dtype to default platform integer
+                    dtype = np.int_
+                else:
+                    dtype = src.dtype
+            else:
+                dtype = out.dtype
+        if (
+            np.issubdtype(src.dtype, np.floating) or np.iscomplexobj(src)
+        ) and np.issubdtype(dtype, np.integer):
+            # Needs changes to convert()
+            raise NotImplementedError(
+                "Integer output types currently not supported for "
+                "floating/complex inputs"
+            )
+        if np.issubdtype(dtype, np.complexfloating) and nan0 is True:
+            # Currently not behaving correctly
+            raise NotImplementedError(
+                "nancumsum and nancumprod currently do not "
+                "supported complex types"
+            )
+        if out is not None:
```

Comment:
suggested edits:
```
if out is not None:
    if dtype != out.dtype:
        # if out array is specified, its type overrules dtype
        dtype = out.dtype
    if axis is not None:
        if out.shape != src.shape:
            raise NotImplementedError(
                "Non-matching shapes on input and output"
                "not supported when axis provided."
            )
    else:
        if out.ndim != 1 or out.size != src.size:
            raise ValueError(
                "Output must be 1D and of the same size as "
                "input when axis not provided"
            )
else:
    out = ndarray(
        shape=src.size if axis is None else src.shape,
        dtype=dtype,
      )


if dtype != src.dtype:
  temp = ndarray(shape=src.shape, dtype=dtype)
  temp._thunk.convert(src._thunk)
  src = temp

out._thunk.scan(op, src._thunk, axis=axis, dtype=dtype, nan0=nan0)
```

Reply:
Some of these are merged into the same error, and the cases where the error should be NotImplemented is separated (the behavior of numpy is very peculiar, we can discuss it if you want).

Reply:
I'm still seeing the input conversion duplicated in the branches. is there any reason that you don't want to factor them out as I suggested?

Reply:
I'd missed the part about factoring and only fixed the error handling issues before! Fixed now.
4bc345e7a91f50a2aa359ea4adbb6f4c1d1cd3de